src/serve.c

```diff
@@ -0,0 +1,157 @@
+#include <assert.h>
+#include <errno.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <arpa/inet.h>
+
+#include "serve.h"
+#include "proto.h"
+#include "util/list.h"
+
+struct server {
+	int sock;
+	list_t *connections;
+	list_t *connect_handlers;
+	list_t *disconnect_handlers;
+};
+
+server_t *start_server(void) {
+	int sock;
+	server_t *server;
+	struct sockaddr_in addr;
+
+	// Open a socket
+	sock = socket(AF_INET, SOCK_STREAM | SOCK_NONBLOCK, 0);
+	if (sock == -1) {
+		perror("Could not open socket");
+		return NULL;
+	}
+
+	// Bind the socket to the address
+	addr.sin_family = AF_INET;
+	addr.sin_port = 1337;
+	addr.sin_addr.s_addr = INADDR_ANY;
+
+	if (bind(sock, (struct sockaddr *) &addr, sizeof(addr)) == -1) {
+		perror("Could not bind socket");
+		return NULL;
+	}
+
+	// Listen on the socket!
+	if (listen(sock, 10) == -1) {
+		perror("Could not listen on socket");
+		return NULL;
+	}
+
+	server = malloc(sizeof(server_t));
+	server->sock = sock;
+	server->connections = list_new();
+	server->connect_handlers = list_new();
+	server->disconnect_handlers = list_new();
+
+	return server;
+}
+
+bool stop_server(server_t *server) {
+	// Close the connection
+	if (close(server->sock) == -1) {
+		perror("Could not close socket");
+		return false;
+	}
+
+	return true;
+}
+
+static void call_handler(connection_handler handler, const connection_t *conn) {
+	handler(conn);
+}
+
+// Called by tick_server upon each new connection found
+static void new_connection(
+    server_t *server,
+    int sock,
+    const struct sockaddr_in *addr,
+    socklen_t len) {
+	// Our handshake response does not change
+	static iiag_handshake_t my_handshake = {
+		IIAG_MARKER,
+		IIAG_PROTO_MAJOR,
+		IIAG_PROTO_MINOR
+	};
+
+	// Recieve the handshake from the client
+	iiag_handshake_t recv_handshake;
+	int ret = recv(sock, &recv_handshake, sizeof(recv_handshake), 0);
```

Comment:
Maybe we should add another abstraction layer that only deals with interacting with the protocol while this only only deals with what to send rather than how to send it? It would make porting to other underlying protocols easier too.